### PR TITLE
Fixes a bug in the blaze/posts endpoint when we get non-200 from WPCOM

### LINKED ISSUE #38030: Blaze unable to handle non-200 api responses
### Impacted plugin

None / Other

### Quick summary

In the `get_blaze_posts()` method, the API response is [handled here](https://github.com/Automattic/jetpack/blob/64517a4d93245a6492c5f65a270c627f1068fdf3/projects/packages/blaze/src/class-dashboard-rest-controller.php#L341-L355). It only checks if the response is a WP_Error. However, in the `request_as_user()` method, the check for [non-200 responses](https://github.com/Automattic/jetpack/blob/64517a4d93245a6492c5f65a270c627f1068fdf3/projects/packages/blaze/src/class-dashboard-rest-controller.php#L829-L831) gets a [WP_REST_Response](https://github.com/Automattic/jetpack/blob/64517a4d93245a6492c5f65a270c627f1068fdf3/projects/packages/blaze/src/class-dashboard-rest-controller.php#L876) in the `get_blaze_error()` method. As a result, this causes a fatal error when the response is not a 200:

```
PHP message: Fatal error: Uncaught Error: Cannot use object of type WP_REST_Response as array in /var/www/wp-content/mu-plugins/jetpack-13.5/jetpack_vendor/automattic/jetpack-blaze/src/class-dashboard-rest-controller.php:351
```

### Steps to reproduce

I'm not able to replicate at will as I'm unsure how to trigger a non-200 response. I came across this issue on plugin installation on helloyumi.com. A thread exists here: p1719257216906619-slack-C06SG526V6X

### A clear and concise description of what you expected to happen.

I expected no PHP fatal messages in the error logs.

### What actually happened

PHP fatal messages in the error logs.

### Impact

Some (< 50%)

### Available workarounds?

No but the platform is still usable

### Platform (Simple and/or Atomic)

Self-hosted

### Logs or notes

VIP

### PR DESCRIPTION
**Pull request by sbarbosa**
Fixes #38030

When we got non-200 for the `/blaze/posts` endpoint, we incorrectly handled an error response in the Jetpack Blaze module. This PR checks if the response is OK to be dealt with and bails early if it's not.

## Proposed changes:

Adds an additional check to control the case that we manually bundle an error inside the Blaze module for that specific call.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p1719257216906619-slack-C06SG526V6X

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

This change is tricky to test because you must generate a non-200 response from the /blaze/posts WPCOM endpoint. 
A code review should be enough, but I will describe the steps in case you have a sandbox to test it.

* Change Jetpack local to use your sandbox
* Modify the WPCOM blaze controller endpoint to return an error (ping me and I can show you the controller location)
* Go to Tools->Advertising
* Verify that you see your error in the network tab. There shouldn't be any fatal error in your site's logs.

### CODE REVIEW

**Review by anomiex — COMMENTED** [review]
You should probably fix this phpdoc comment
https://github.com/Automattic/jetpack/blob/80b607b665776229191c298bfe0b11f5338113c6/projects/packages/blaze/src/class-dashboard-rest-controller.php#L852-L859
It looks like that will only ever return a `WP_REST_Response`, never a `WP_Error`.

Phan is flagging that as `PhanTypeMismatchReturnProbablyReal`, BTW. It might be worth going through the rest of the Phan issues in the package (e.g. run `jetpack phan --no-baseline packages/blaze`) looking for similar issues.

**Comment by sbarbosa** [comment]
@anomiex, thanks for pointing me to how to test the static analysis locally!
I fixed all of the errors in my last commit.

File: projects/packages/blaze/.phan/baseline.php [hunk 1]
```diff
@@ -8,19 +8,9 @@
...
 return [
-    // # Issue statistics:
-    // PhanTypeMismatchArgument : 2 occurrences
-    // PhanTypeMismatchReturn : 2 occurrences
-    // PhanParamTooMany : 1 occurrence
-    // PhanTypeMismatchArgumentProbablyReal : 1 occurrence
-    // PhanTypeMismatchReturnProbablyReal : 1 occurrence
-    // PhanTypeNoAccessiblePropertiesForeach : 1 occurrence
-
+    // This baseline has no suppressions
```
Comment (anomiex) [inline projects/packages/blaze/.phan/baseline.php:11]: :tada: Thanks!

File: projects/packages/blaze/src/class-blaze.php [hunk 2]
```diff
@@ -123,7 +123,7 @@ public static function enable_blaze_menu() {
 				__( 'Advertising', 'jetpack-blaze' ),
 				'manage_options',
 				'https://wordpress.com/advertising/' . $domain,
-				null,
+				null, // @phan-suppress-current-line PhanTypeMismatchArgumentProbablyReal -- Core should ideally document null for no-callback arg. TODO add link with Trac issue.
```
Comment (anomiex) [inline projects/packages/blaze/src/class-blaze.php]: I found an existing trac issue for this. 🙂
```suggestion
				null, // @phan-suppress-current-line PhanTypeMismatchArgumentProbablyReal -- Core should ideally document null for no-callback arg. https://core.trac.wordpress.org/ticket/52539
```

Comment (anomiex) [reply projects/packages/blaze/src/class-blaze.php]: I created https://github.com/Automattic/jetpack/pull/38075 to update the other instances of this comment currently in the monorepo.

**Comment by timur987** [comment]
@sbarbosa, I looked through the code and it works fine. Stumbled when started testing, but will keep trying a bit longer.

image

Text says: `Cannot find module './en-us'` did you encountered it before?


----

Update:
I tried to use my A8C email and ignored Webpack's errors. The issue now is connected to getting posts: `Blaze is syncing your content as part of first-time setup – this can take up to 15 minutes or a few hours.`

image

**Comment by sbarbosa** [comment]
Hi @timur987, I'm not familiar with the first error you shared. It never happened to me when loading Jetpack locally.

The second error is expected if your site is syncing content to WPCOM. It should disappear after a few minutes. 
I had some problems with the connection for my Jetpack site, and I saw that message in the dashboard. I checked the Jetapck Debug page for my site, and it had some warnings and problems with the sync, so I had to restore my local.

I am not sure if that happened to you too, but this bug doesn't relate to that message. You can change [this function](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/blaze/src/class-dashboard-rest-controller.php#L908-L915) to return true so you can fully test my change.

Let me tag @diogoca, just in case he knows a way to check if there is a problem with the sync of your site, or if there are some edge cases that we need to be aware for the `are_posts_ready` function.

**Comment by diogoca** [comment]
> Update: I tried to use my A8C email and ignored Webpack's errors. The issue now is connected to getting posts: `Blaze is syncing your content as part of first-time setup – this can take up to 15 minutes or a few hours.`
> 
> image

This is expected after you connect your site to Jetpack the first time; if the message won't disappear after 10-15 minutes, there should be an issue with your sync; we may confirm the Health in the debugger: Pc9OEs-v-p2

image
@timur987 I'm not sure if this is happening because your tube is offline, but it looks like there is an issue with the full sync. Could you please try to debug again when your tunnel is up to see if this is still working?

**Comment by timur987** [comment]
(Updated this comment after additional testing)

Thanks for jumping in @diogoca! I checked again today (previously tried yesterday) and nothing has changed. By the way, I used a piece of advise from @sbarbosa and just returned true and it helped.

> You can change this function to return true so you can fully test my change.

The fix was in returning `true` in `are_posts_ready` function.

---

In order to broke /blaze/posts request, in `wp-content/rest-api-plugins/endpoints/blaze.php`, the result was:

image

---

Then I switched to this branch and checked the same. With the new changes it just returned the WPCOM's response without throwing an error:

image

**Review by timur987 — APPROVED** [review]
LGTM and works as advertised! 👍

**Pull request merged**